talisker/context.py

```diff
@@ -0,0 +1,136 @@
+# Copyright (C) 2016- Canonical Ltd
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+
+from __future__ import unicode_literals
+from __future__ import print_function
+from __future__ import division
+from __future__ import absolute_import
+
+from builtins import *  # noqa
+
+try:
+    from collections.abc import Mapping
+except:  # < py3.3
+    from collections import Mapping
+
+from collections import OrderedDict
+from contextlib import contextmanager
+
+from werkzeug.local import Local, LocalManager, release_local
+
+# a per request/job context. Generally, this will be the equivalent of thread
+# local storage, but if greenlets are being used it will be a greenlet local.
+context = Local()
+manager = LocalManager(context)
+
+
+def clear():
+    release_local(context)
+
+
+class ContextStack(Mapping):
+    """A stacked set of dicts stored in a context.
+
+    Support lookups and iteration, which go from top of the stack down.
+    Can also be used as a context manager.
+
+    """
+
+    def __init__(self, name, *dicts):
+        """Initialise stack, with name to use in context storage."""
+        self.name = name
+        self._stack.extend(dicts)
+        self._flat = None
+
+    @property
+    def _stack(self):
+        stack = getattr(context, self.name, None)
+        if stack is None:
+            stack = []
+            setattr(context, self.name, stack)
+            self._clear()
```

Comment:
No, the context was not initialised.

I've renamed to self._clear_flat, to avoid any confusion - I'm just invalidating the flattened cache, not resetting the list.